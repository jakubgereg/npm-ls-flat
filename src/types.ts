export type DependencyList = Record<string, string>;

export interface PackageJsonType {
  dependencies?: DependencyList;
  devDependencies?: DependencyList;
}

export type DependencyTree = Record<string, PackageType>;

export interface PackageType {
  version: string;
  invalid?: string;
  dependencies?: DependencyTree;
}

export interface PackageBaseInfo {
  name: string;
  version: string;
}

export interface PackageInfo extends PackageBaseInfo {
  name: string;
  version: string;
  invalid?: string;
  path?: PackageBaseInfo[];
}
