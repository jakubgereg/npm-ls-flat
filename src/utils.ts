import semver from 'semver';
import type { DependencyTree, PackageInfo, PackageType } from './types';

export const isPackageType = (obj: any): obj is PackageType =>
  typeof obj === 'object' && typeof obj.version === 'string' && semver.valid(obj.version) !== null;

export const hasDependencies = (obj: any): obj is PackageType & { dependencies: DependencyTree } =>
  typeof obj === 'object' && typeof obj.dependencies === 'object';

const transformToPackageInfo = (pkg: PackageType, path: string[] = []): PackageInfo[] => {
  const result: PackageInfo[] = [];
  if (isPackageType(pkg)) {
    result.push({
      name: path[path.length - 1], // Get the name from the last element of the path
      version: pkg.version,
      invalid: pkg.invalid,
      path: path.length === 1 ? undefined : path.slice(0, -1) // Set to undefined if path is empty
    });
  }
  if (hasDependencies(pkg)) {
    for (const [key, value] of Object.entries(pkg.dependencies)) {
      result.push(...transformToPackageInfo(value, [...path, key]));
    }
  }
  return result;
};

export const traverseDependencyTree = (dependencyTree: DependencyTree): PackageInfo[] => {
  const result: PackageInfo[] = [];
  for (const [name, pkg] of Object.entries(dependencyTree)) {
    result.push(...transformToPackageInfo(pkg, [name]));
  }
  return result;
};
