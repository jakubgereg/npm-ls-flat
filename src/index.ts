#!/usr/bin/env node

import { exec } from 'node:child_process';
import path from 'node:path';
import chalk from 'chalk';
import fs from 'fs-extra';
import lodash from 'lodash';
import semver from 'semver';

import type { DependencyList, DependencyTree, PackageInfo, PackageJsonType, PackageType } from './types.js';
import { sortPackageVersions, traverseDependencyTree } from './utils.js';

const readPackageJsonFile = async (path: string, allowDev = false): Promise<DependencyList> => {
  try {
    const { dependencies, devDependencies }: PackageJsonType = await fs.readJson(path);
    let allPackages = { ...dependencies };
    if (allowDev && devDependencies) allPackages = { ...allPackages, ...devDependencies };
    return allPackages;
  } catch (err) {
    console.error('Error reading package.json:', err);
    throw err;
  }
};

const getDependencyTree = (packages: string[] = []): Promise<DependencyTree | undefined> =>
  new Promise((resolve, reject) => {
    const npmls = packages.length ? `npm ls ${packages.join(' ')} --all --json` : 'npm ls --all --json';
    exec(npmls, (_, stdout) => {
      try {
        const { dependencies } = JSON.parse(stdout) as PackageType;
        resolve(dependencies);
      } catch (parseError) {
        reject(parseError);
      }
    });
  });

const findMismatchedVersions = (deps: Record<string, PackageInfo[]>): Record<string, [PackageInfo, PackageInfo[]]> =>
  Object.entries(deps).reduce(
    (acc, [name, dependencies]) => {
      const root = dependencies.find((dependency) => !dependency.path);
      if (!root) return acc;
      const differentVersions = dependencies.filter((dependency) => dependency.version !== root.version);
      const versions: [PackageInfo, PackageInfo[]] | [] = differentVersions.length
        ? [root, dependencies.filter((dependency) => dependency.version !== root.version)]
        : [root, []];

      if (versions[1].length > 0 || root.invalid) acc[name] = versions;
      return acc;
    },
    {} as Record<string, [PackageInfo, PackageInfo[]]>
  );

(async () => {
  const packageJsonPath = path.resolve(process.cwd(), 'package.json');
  try {
    const packageNames = Object.keys(await readPackageJsonFile(packageJsonPath, true));
    console.log(
      `Looking for mismatches in ${chalk.greenBright(packageNames.length)} ${
        packageNames.length > 1 ? 'packages' : 'package'
      }...`
    );
    const dependencyTree = await getDependencyTree(packageNames);
    const groupedDependencyTree = lodash.groupBy(
      traverseDependencyTree(dependencyTree ?? {}).filter((dependency) =>
        packageNames.some((name) => dependency.name === name)
      ),
      'name'
    );

    const mismatches = Object.values(findMismatchedVersions(groupedDependencyTree))
      //sort first packages that have invalid versions
      .sort((a) => (a[0].invalid ? -1 : 1));

    if (!mismatches.length) {
      console.log(chalk.green('All packages have consistent versions.'));
      return;
    }

    for (const [root, pckgs] of mismatches) {
      console.log(
        `\n${root.name} ${chalk.cyanBright(root.version)} ${root.invalid ? chalk.redBright('(invalid)') : ''}`
      );
      if (root.invalid) console.log(` ${chalk.redBright(root.invalid)}`);

      const sortedPackages = sortPackageVersions(pckgs, 'DESC');

      const packageVersionColor = (version: string) =>
        semver.gt(root.version, version) ? chalk.redBright(version) : chalk.greenBright(version);
      for (const pckg of sortedPackages) {
        console.log(
          ` ${packageVersionColor(pckg.version)} (${chalk.yellow(pckg.path?.map(({ name }) => name)?.join(' > '))})`
        );
      }
    }
  } catch (err) {
    console.error('Failed to get packages:', err);
  }
})();
