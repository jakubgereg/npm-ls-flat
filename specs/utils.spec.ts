import { DependencyTree } from '../src/types';
import { hasDependencies, isPackageType, traverseDependencyTree } from '../src/utils';

const dependencyTree: DependencyTree = {
  'package-a': {
    version: '1.0.0',
    dependencies: {
      'package-b': {
        version: '1.1.0',
        dependencies: {
          'package-c': {
            version: '1.2.0'
          }
        }
      }
    }
  },
  'package-x': {
    version: '2.0.0',
    dependencies: {
      'package-y': {
        version: '3.0.0',
        dependencies: {
          'package-z': {
            version: '4.0.0'
          }
        }
      }
    }
  },
  'package-c': {
    version: '1.2.0'
  },
  'package-d': {
    version: '1.3.0',
    dependencies: {
      'package-a': {
        version: '1.0.1'
      }
    }
  }
};

const brokenDependencyTree: any = {
  'package-a': {
    version: '_invalid_version_'
  },
  'package-b': {
    version: '1.1.0',
    dependencies: ['package-1', 'package-2', 'package-3']
  },
  'package-c': {
    version: '1.2.0',
    dependencies: {
      'package-d': {
        version: '_invalid_version_2'
      },
      'package-e': {
        version: 0
      },
      'package-f': {
        version: '1.0.122'
      }
    }
  }
};

describe('traverseDependencyTree', () => {
  it('should return an empty array', () => {
    expect(traverseDependencyTree({})).toEqual([]);
  });
  it('should traverse dependency tree', () => {
    const result = traverseDependencyTree(dependencyTree);
    expect(result).toEqual([
      { name: 'package-a', version: '1.0.0', path: undefined },
      { name: 'package-b', version: '1.1.0', path: [{ name: 'package-a', version: '1.0.0' }] },
      {
        name: 'package-c',
        version: '1.2.0',
        path: [
          { name: 'package-a', version: '1.0.0' },
          { name: 'package-b', version: '1.1.0' }
        ]
      },
      { name: 'package-x', version: '2.0.0', path: undefined },
      { name: 'package-y', version: '3.0.0', path: [{ name: 'package-x', version: '2.0.0' }] },
      {
        name: 'package-z',
        version: '4.0.0',
        path: [
          { name: 'package-x', version: '2.0.0' },
          { name: 'package-y', version: '3.0.0' }
        ]
      },
      { name: 'package-c', version: '1.2.0', path: undefined },
      { name: 'package-d', version: '1.3.0', path: undefined },
      { name: 'package-a', version: '1.0.1', path: [{ name: 'package-d', version: '1.3.0' }] }
    ]);
    expect(traverseDependencyTree(dependencyTree)).toMatchSnapshot();
  });

  it('should handle broken dependency tree', () => {
    const result = traverseDependencyTree(brokenDependencyTree);
    expect(result).toEqual([
      { name: 'package-b', version: '1.1.0', path: undefined },
      { name: 'package-c', version: '1.2.0', path: undefined },
      { name: 'package-f', version: '1.0.122', path: [{ name: 'package-c', version: '1.2.0' }] }
    ]);
    expect(traverseDependencyTree(brokenDependencyTree)).toMatchSnapshot();
  });

  it('should match snapshot', () => {});
});

describe('isPackageType', () => {
  it('should return true', () => {
    expect(isPackageType({ version: '1.2.3' })).toBe(true);
  });
  it('should return false', () => {
    expect(isPackageType({ name: 'my package' })).toBe(false);
    expect(isPackageType(['1.2.3'])).toBe(false);
    expect(isPackageType({ version: 3 })).toBe(false);
    expect(isPackageType({ version: '_ivalid_version_' })).toBe(false);
  });
});

describe('hasDependencies', () => {
  it('should return true', () => {
    expect(hasDependencies({ version: '1.2.3', dependencies: {} })).toBe(true);
  });
  it('should return false', () => {
    expect(hasDependencies({ name: 'test', dependencies: undefined })).toBe(false);
    expect(hasDependencies({ dependencies: 'someDeps' })).toBe(false);
  });
});
