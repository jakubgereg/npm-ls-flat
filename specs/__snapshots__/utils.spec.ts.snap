// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traverseDependencyTree should handle broken dependency tree 1`] = `
[
  {
    "invalid": undefined,
    "name": "package-b",
    "path": undefined,
    "version": "1.1.0",
  },
  {
    "invalid": undefined,
    "name": "package-c",
    "path": undefined,
    "version": "1.2.0",
  },
  {
    "invalid": undefined,
    "name": "package-f",
    "path": [
      {
        "name": "package-c",
        "version": "1.2.0",
      },
    ],
    "version": "1.0.122",
  },
]
`;

exports[`traverseDependencyTree should traverse dependency tree 1`] = `
[
  {
    "invalid": undefined,
    "name": "package-a",
    "path": undefined,
    "version": "1.0.0",
  },
  {
    "invalid": undefined,
    "name": "package-b",
    "path": [
      {
        "name": "package-a",
        "version": "1.0.0",
      },
    ],
    "version": "1.1.0",
  },
  {
    "invalid": undefined,
    "name": "package-c",
    "path": [
      {
        "name": "package-a",
        "version": "1.0.0",
      },
      {
        "name": "package-b",
        "version": "1.1.0",
      },
    ],
    "version": "1.2.0",
  },
  {
    "invalid": undefined,
    "name": "package-x",
    "path": undefined,
    "version": "2.0.0",
  },
  {
    "invalid": undefined,
    "name": "package-y",
    "path": [
      {
        "name": "package-x",
        "version": "2.0.0",
      },
    ],
    "version": "3.0.0",
  },
  {
    "invalid": undefined,
    "name": "package-z",
    "path": [
      {
        "name": "package-x",
        "version": "2.0.0",
      },
      {
        "name": "package-y",
        "version": "3.0.0",
      },
    ],
    "version": "4.0.0",
  },
  {
    "invalid": undefined,
    "name": "package-c",
    "path": undefined,
    "version": "1.2.0",
  },
  {
    "invalid": undefined,
    "name": "package-d",
    "path": undefined,
    "version": "1.3.0",
  },
  {
    "invalid": undefined,
    "name": "package-a",
    "path": [
      {
        "name": "package-d",
        "version": "1.3.0",
      },
    ],
    "version": "1.0.1",
  },
]
`;
